tests/jballerina-unit-test/src/test/resources/test-src/expressions/varref/record-variable-reference.bal

```diff
@@ -369,7 +519,110 @@ function testMappingBindingWithSingleNameFieldBinding() {
+
+    [record {|int a; int b;|}] r3 = [{a: 3, b: 4}];
+    [{a, b}] = [...r3];
+    assertEquality(3, a);
+    assertEquality(4, b);
+
+    record {|int b; string c;|} rec1 = {b: 2, c: "C"};
+    string c;
+    [{a, b, c}] = [{a: 1, ...rec1}];
+    assertEquality(1, a);
+    assertEquality(2, b);
+    assertEquality("C", c);
+
+    [record {int a; int b; stream<int> c;}] r4 = [{a: 5, b: 6, c: new}];
+    [{a, b}] = r4;
+    assertEquality(5, a);
+    assertEquality(6, b);
+
+    string fname;
+    string lname;
+    string id;
+    int age;
+    string nextId;
+    [record {string id; string fname; string lname;}, record {|string id; int age;|}] r5
+                                    = [{id: "u1001", fname: "John", lname: "doe"}, {id: "u1002", age: 10}];
+    [{id, fname, lname}, {id: nextId, age}] = r5;
+    assertEquality("u1001", id);
+    assertEquality("John", fname);
+    assertEquality("doe", lname);
+    assertEquality("u1002", nextId);
+    assertEquality(10, age);
+
+    int p1;
+    int p2;
+    int p3;
+    int p4;
+    int p5;
+    [[record {int p1; int p2;}], [[record {int p3; int p4;}], record {int p5;}]] r6
+                                    = [[{p1: 10, p2: 12}], [[{p3: 13, p4: 14}], {p5: 15}]];
+    [[{p1, p2}], [[{p3, p4}], {p5}]] = r6;
+    assertEquality(10, p1);
+    assertEquality(12, p2);
+    assertEquality(13, p3);
+    assertEquality(14, p4);
+    assertEquality(15, p5);
+
+    [record {|int a; int b; anydata...;|}] r7 = [{a: 1, b: 2}];
+    [{a, b}] = r7;
+    assertEquality(1, a);
+    assertEquality(2, b);
+
+    [record {|int a; int b; anydata...;|}] r8 = [{a: 10, b: 12}];
+    [{a, b}] = [...r8];
+    assertEquality(10, a);
+    assertEquality(12, b);
+
+    int x1;
+    int x2;
+    var getMap = function() returns record {record {int x1; int x2;} x3;} => {x3: {x1: 2, x2: 3}};
+    {x3: {x1, x2}} = getMap();
+    assertEquality(2, x1);
+    assertEquality(3, x2);
+    assertEquality({x1: 2, x2: 3}, {x1, x2});
+
+    // TODO: Uncomment after fixing #40312
+    // int d;
+    // error<record { record { int d; } x; }> err = error("Transaction Failure", x = {d: 0});
+
+    // error(x = {d}) = err;
+}
+
+function testRecordTypedBindingPatternAgainstRecordLiteralInRecordDestructuring() {
+    record {|string state; int zipCode;|} {state, zipCode} = {state: "Alabama", zipCode: 35004};
+    assertEquality("Alabama", state);
+    assertEquality(35004, zipCode);
+
+    record {|string street; float...;|} {street, ...coordinates} = {street: "Highway 61", "lat": 37.30, "lon": -90.40};
+    assertEquality("Highway 61", street);
+    assertEquality({lat: 37.30, lon: -90.40}, coordinates);
+
+    record {string state;} {state: state2, ...other} = {state: "Colorado", "zipCode": 80001};
+    assertEquality("Colorado", state2);
+    assertEquality({zipCode: 80001}, other);
+
+    var {country, ...rest} = {country: "USA", zipCode: 80001};
+    assertEquality("USA", country);
+    assertEquality({zipCode: 80001}, rest);
```

Comment:
Add more tests including where the member binding patterns are in turn structured binding patterns.

Also add negative tests for unspecified fields, optional field binding, etc.

Reply:
Added in https://github.com/ballerina-platform/ballerina-lang/pull/40283/commits/edebb78afebb8f6e60721465d50959820908d968